jdaviz/core/template_mixin.py

```diff
@@ -3127,20 +3141,137 @@ class Plot(PluginSubcomponent):
+            return
+        data = self.app.data_collection[label]
+
+        # if not provided, fallback on existing data
+        length_mismatch = False
+        for component in self._viewer_components:
+            kwargs.setdefault(component, data[component])
+            if len(kwargs[component]) != len(data[component]):
+                length_mismatch = True
+
+        if not length_mismatch:
+            # then we can update the existing entry
+            components = {c.label: c for c in data.components}
+            data.update_components({components[comp]: kwargs[comp]
+                                    for comp in self._viewer_components})
+        else:
+            # then we need to replace the existing entry, restoring any existing styles,
+            # if they exist
+            if label in self.layers.keys():
+                style_state = self.layers[label].state.as_dict()
+            else:
+                style_state = {}
+            self._remove_data(label)
+            self._add_data(label, **kwargs)
+            self.update_style(label, **style_state)
+        if reset_lims:
+            self.viewer.state.reset_limits()
+
+    def update_style(self, label, **kwargs):
+        kwargs.setdefault('visible', True)
+        if label not in self.layers.keys():
+            if not kwargs['visible']:
+                # then we were only trying to hide anyways
+                # (note: this ignores any other passed styles)
+                return
+        dc_entry = self.app.data_collection[label]
+        if kwargs['visible']:
+            if label not in self.layers.keys():
+                self.viewer.add_data(dc_entry)
+        else:
+            # remove from viewer, leave in app (note: this will clear style options)
+            # NOTE: if we want to keep styles, we could skip this and only toggle visibilities,
+            # but then the zooming logic will need to be updated to account for visibility
+            # states and the if not kwargs['visible'] check above to return should ensure no
+            # style options are passed
+            self.viewer.remove_data(dc_entry)
+            return
+
+        lyr = self.layers[label]
+        with delay_callback(lyr.state, *list(kwargs.keys())):
+            for k, v in kwargs.items():
+                if k == 'layer' or k.endswith('_att'):
+                    continue
+                setattr(lyr.state, k, v)
+
+    def _add_data(self, label, **kwargs):
+        self._check_valid_components(**kwargs)
+        data = Data(label=label, **kwargs)
+        dc = self.app.data_collection
+        dc.append(data)
+        dc_entry = dc[label]
+
+        if len(dc) > 1:
+            # we can assume the same units/components since this only accepts x and y
+            ref_data = dc[0]
+            links = [LinkSame(dc_entry.components[i], ref_data.components[i])
+                     for i in range(1, len(ref_data.components))]
+            dc.add_link(links)
```

Comment:
I vaguely remember @astrofrog said we should do something like this? I see it used in Imviz helper.

```suggestion
            with self.app.data_collection.delay_link_manager_update():
                dc.add_link(links)
```

Reply:
does this do anything if there is only a _single_ call to `add_link`?

Reply:
Hmm maybe not...? https://jdaviz.readthedocs.io/en/latest/dev/links.html#speeding-up-adding-links-to-the-data-collection